packages/shared/src/hooks/notifications/useNotificationPreference.ts

```diff
@@ -0,0 +1,113 @@
+import { Squad } from '../../graphql/sources';
+import { updateFlagsCache } from '../../graphql/source/common';
+
+interface UseNotificationPreference {
+  isFetching: boolean;
+  isPreferencesReady: boolean;
+  preferences: NotificationPreference[];
+  hideSourceFeedPosts(): Promise<unknown>;
+  showSourceFeedPosts(): Promise<unknown>;
+  muteNotification: typeof muteNotification;
+  clearNotificationPreference: typeof clearNotificationPreference;
+}
+
+interface UseNotificationPreferenceProps {
+  params: Parameters<typeof getNotificationPreferences>[0];
+  squad?: Squad;
+}
+
+export const checkHasMutedPreference = (
+  { notificationType, referenceId, status }: NotificationPreference,
+  type: NotificationType,
+  id: string,
+): boolean =>
+  status === NotificationPreferenceStatus.Muted &&
+  notificationType === type &&
+  referenceId === id;
+
+export const useNotificationPreference = ({
+  params,
+  squad,
+}: UseNotificationPreferenceProps): UseNotificationPreference => {
+  const { user } = useAuthContext();
+  const client = useQueryClient();
+  const key = generateQueryKey(RequestKey.NotificationPreference, user, params);
+  const { data, isFetched, isLoading } = useQuery(
+    key,
+    () => getNotificationPreferences(params),
+    { enabled: params?.length > 0 },
+  );
+  const { mutateAsync: muteNotificationAsync } = useMutation(muteNotification, {
+    onSuccess: (_, { referenceId, type }) => {
+      client.setQueryData<NotificationPreference[]>(key, (oldData = []) => {
+        const preference: NotificationPreference = {
+          referenceId,
+          notificationType: type,
+          type: notificationPreferenceMap[type],
+          userId: user.id,
+          status: NotificationPreferenceStatus.Muted,
+        };
+
+        return [...oldData, preference];
+      });
+    },
+  });
+
+  const { mutateAsync: clearNotificationPreferenceAsync } = useMutation(
+    clearNotificationPreference,
+    {
+      onSuccess: (_, { referenceId, type }) => {
+        client.setQueryData<NotificationPreference[]>(key, (oldData) => {
+          if (!oldData) return [];
+
+          return oldData.filter(
+            (preference) =>
+              !checkHasMutedPreference(preference, type, referenceId),
+          );
+        });
+      },
+    },
+  );
+
+  const { mutateAsync: hideSourceFeedPostsAsync } = useMutation(
+    hideSourceFeedPosts,
+    {
+      onSuccess: () => {
+        updateFlagsCache(client, squad, user, { hideFeedPosts: true });
+      },
+    },
+  );
+
+  const { mutateAsync: showSourceFeedPostsAsync } = useMutation(
+    showSourceFeedPosts,
+    {
+      onSuccess: () => {
+        updateFlagsCache(client, squad, user, { hideFeedPosts: false });
+      },
+    },
+  );
+
+  return {
+    hideSourceFeedPosts: () => hideSourceFeedPostsAsync(squad?.id),
+    showSourceFeedPosts: () => showSourceFeedPostsAsync(squad?.id),
+    isFetching: isLoading,
+    preferences: data ?? [],
```

Comment:
You can move `[]` to `initialData: () => []` in `useQuery`.

Reply:
Oh, right. That is much better.

Reply:
This is an outdated comment, kindly check the files tab for the most updated ones.

Reply:
Oh dang, for some reason it kept showing the old one, I see it now!
Sorry github must be tripping on me